libraries/core_libs/network/src/tarcap/packets_handlers/transaction_packet_handler.cpp

```diff
@@ -17,36 +17,57 @@ TransactionPacketHandler::TransactionPacketHandler(std::shared_ptr<PeersState> p
 
 void TransactionPacketHandler::validatePacketRlpFormat(const PacketData &packet_data) const {
   auto items = packet_data.rlp_.itemCount();
-  if (items == 0 || items > kMaxTransactionsInPacket) {
-    throw InvalidRlpItemsCountException(packet_data.type_str_, items, kMaxTransactionsInPacket);
+  if (items != kTransactionPacketItemCount) {
+    throw InvalidRlpItemsCountException(packet_data.type_str_, items, kTransactionPacketItemCount);
+  }
+  auto hashes_count = packet_data.rlp_[0].itemCount();
+  auto trx_count = packet_data.rlp_[1].itemCount();
+
+  if (hashes_count != trx_count) {
+    throw InvalidRlpItemsCountException(packet_data.type_str_, hashes_count, trx_count);
+  }
+  if (hashes_count == 0 || hashes_count > kMaxTransactionsInPacket) {
+    throw InvalidRlpItemsCountException(packet_data.type_str_, hashes_count, kMaxTransactionsInPacket);
```

Comment:
So I find the rlp structure of TransactionPacket little bit weird. Why dont we simply have array of pairs<tx_hash, tx_object> so it would look like this:
[
[tx_hash, [tx_obj_var1, tx_obj_var2, ....]]
]

So when you validate it, you check the size of rlp to be <= kMaxTransactionsInPacket as before
Then when you start parsing insice packet handler, you iterate over the outer rlp list and for each item(tx pair), you check if tx_rlp.itemsCount() == 2 and then you parse only tx hash (tx_rlp[0]), check it against caches etc... and if it is not known you will parse actual object tx_rlp[1] and process it

Reply:
so the reason why I went for this structure is that it is costly to iterate back and forth over rlp, and it is common that we get a lot of transactions more than once so in many cases I will only be looking for transactions hashes and not the full transactions, so it is fastest to process the first chunk of only hashes first and go to the second chunk where transactions are only for those transactions that we need

Reply:
rlp does not have random access, it iterates to get to the position it needs

Reply:
It would look like this:
```
for (size_t tx_idx = 0; tx_idx < packet_data.rlp_.itemCount(); tx_idx++) {
  const auto& tx_rlp = packet_data.rlp_[tx_idx];
  if (tx_rlp.itemCount() != 2) {
  	throw MaliciousPeerException(msg);
  }
  
  const auto tx_hash = tx_rlp[0].toHash();
  
  // Do all the checks with tx hash, e.g.
  if (isKnownTx(tx_hash)) {
    continue;
  }
  
  const auto tx_obj = std::make_shared<Transaction>(tx_rlp_[1]);

  // Process tx obj
}
```

Reply:
In effort to reduce unnecessary gossip of transactions my goal was to mark transaction as known as soon as we receive it from the peer. So if we receive 1000 transactions, I want to mark these transactions as known for the peer we received it from because there is a great chance that a lot of these transactions we already have and we are about to send it to the same peer we just received it from. If I were to do it in the loop you proposed, deserializing transactions would slow this down. But if I were to use your format and only get hashes in one pass and then transactions in another pass that would be slower. This is what I just measured and for 1000 transactions and doing two passes the difference is this:
Elapsed method 1 : 955 microseconds
Elapsed method 2 : 1762 microseconds

But which ever format we choose I would definitely recommend first marking peer transction as known before doing the deserialization of transactions.

Reply:
if needed we can address this in a separate PR, resolving to merge this PR

Reply:
@mfrankovi  yeah so we do send votes periodically every 100ms, we do not gossip it with each incoming tx packet. So if the difference between these 2 methods is 800us, I think it is not worth it and we should choose the approach, which is better from design/flow point of view because saving 800us will have almost zero effect on marking txs as known if we send it txs every 100ms....